flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandlerTest.java

```diff
@@ -0,0 +1,207 @@
+import static org.hamcrest.CoreMatchers.not;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.empty;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for {@link StreamOperatorStateHandlerTest}.
+ */
+public class StreamOperatorStateHandlerTest {
+	/**
+	 * Tests that a failing snapshot method call to the keyed state backend will trigger the closing
+	 * of the StateSnapshotContextSynchronousImpl and the cancellation of the
+	 * OperatorSnapshotResult. The latter is supposed to also cancel all assigned futures.
+	 */
+	@Test
+	public void testFailingBackendSnapshotMethod() throws Exception {
+		final long checkpointId = 42L;
+		final long timestamp = 1L;
+
+		final CloseableRegistry closeableRegistry = new CloseableRegistry();
+
+		RunnableFuture<SnapshotResult<KeyedStateHandle>> keyedStateManagedFuture = new CancelableFuture<>();
+		RunnableFuture<SnapshotResult<KeyedStateHandle>> keyedStateRawFuture = new CancelableFuture<>();
+		RunnableFuture<SnapshotResult<OperatorStateHandle>> operatorStateManagedFuture = new CancelableFuture<>();
+		RunnableFuture<SnapshotResult<OperatorStateHandle>> operatorStateRawFuture = new CancelableFuture<>();
+
+		OperatorSnapshotFutures operatorSnapshotResult = new OperatorSnapshotFutures(
+			keyedStateManagedFuture,
+			keyedStateRawFuture,
+			operatorStateManagedFuture,
+			operatorStateRawFuture);
+
+		StateSnapshotContextSynchronousImpl context = new TestStateSnapshotContextSynchronousImpl(checkpointId, timestamp, closeableRegistry);
+		context.getRawKeyedOperatorStateOutput();
+		context.getRawOperatorStateOutput();
+
+		StreamTaskStateInitializerImpl stateInitializer =
+			new StreamTaskStateInitializerImpl(new MockEnvironmentBuilder().build(), new MemoryStateBackend());
+		StreamOperatorStateContext stateContext = stateInitializer.streamOperatorStateContext(
+			new OperatorID(),
+			"whatever",
+			new TestProcessingTimeService(),
+			new UnUsedKeyContext(),
+			IntSerializer.INSTANCE,
+			closeableRegistry,
+			new InterceptingOperatorMetricGroup());
+		StreamOperatorStateHandler stateHandler = new StreamOperatorStateHandler(stateContext, new ExecutionConfig(), closeableRegistry);
+
+		final String keyedStateField = "keyedStateField";
+		final String operatorStateField = "operatorStateField";
+
+		stateHandler.setCurrentKey(44L);
+		stateHandler.initializeOperatorState(stateSnapshotContext -> {
+			stateSnapshotContext.getKeyedStateStore()
+				.getState(new ValueStateDescriptor<>(keyedStateField, LongSerializer.INSTANCE))
+				.update(42L);
+			stateSnapshotContext.getOperatorStateStore()
+				.getListState(new ListStateDescriptor<>(operatorStateField, LongSerializer.INSTANCE))
+				.add(42L);
+		});
+
+		assertThat(stateContext.operatorStateBackend().getRegisteredStateNames(), is(not(empty())));
+		assertThat(stateContext.keyedStateBackend().numKeyValueStatesByName(), equalTo(1));
+
+		try {
+			stateHandler.snapshotState(
```

Comment:
Looks like two test cases to me: in this test, I'd actually would like to see a successful snapshot. And then have second with failure.